Add ENABLE_PUSH flag in the Upgrade HTTP2-Settings header

hkwi opened the pull request:
Push flag required for the case the initial upgrade request triggered server push. This PR is a relaxed version of #310 , which was more strict on server push I/O timing.
Test send/recv timing was fixed, because connection state machine does not get corrupted during h2c upgrade migration.

hkwi committed:
Add ENABLE_PUSH flag in the Upgrade HTTP2-Settings header

Push flag required for the case the initial upgrade request triggered server push.

hkwi committed:
fix test synchronization

Some tests requires synchronization to let the connection
state machine working.

Lukasa commented on test/test_hyper.py:
```diff
@@ -1303,6 +1303,188 @@ def test_resetting_streams_after_close(self):
+        frame.data = data
+        if end_stream:
+            frame.flags.add('END_STREAM')
+        self.frames.append(frame)
```
Can we move these helpers out to a common mixin class now that we're using them in more than one place?

hkwi replied:
I got it.

Lukasa commented on test/test_hyper.py:
```diff
@@ -1303,6 +1303,188 @@ def test_resetting_streams_after_close(self):
+            self.http101 + b''.join([frame.serialize()
+                                     for frame in self.frames])
+        )
+        self.conn.request('GET', '/')
```
I wonder if this should do any asserting that the Upgrade header is present?

hkwi replied:
Upgrading process itself is tested at `test_http11.py`, and IMHO, we would focus on client side view here.

Lukasa replied:
Yeah, that's fair enough.

Lukasa commented on test/test_integration.py:
```diff
@@ -958,16 +1004,15 @@ def socket_handler(listener):
             sock.sendall(f.serialize())
-
-            # Wait for the message from the main thread.
-            send_event.wait()
```
Why did this get moved?

hkwi replied:
We have to wait here only when the client side are going to get the response. Another option is add `send_event.wait()` here, and add `get_response()`, `send_event.set()` in client side.

Lukasa replied:
So historically the reason for these waits has been to ensure that we don't close sockets in a way that leads to unexpected socket behaviour on some platforms. If we don't *need* to remove them I'd be inclined to leave them in.

hkwi replied:
OK

Lukasa commented on test/test_integration.py:
```diff
@@ -1078,8 +1126,6 @@ def socket_handler(listener):
             d.flags.add('END_STREAM')
             sock.send(d.serialize())
-
-            send_event.wait(5)
```
Why was this removed?

hkwi replied:
We have to wait here when we close the socket. Another option is add `send_event.wait()` and `sock.close()` here, and `send_event.set()` in client side.

Lukasa replied:
I don't think this really answered what the reason was for removing this. Was it causing a problem?

hkwi replied:
It does not causing problem. I removed just because I thought it was not necessary.

Lukasa replied:
Yeah, for this stuff that's not really a great attitude. Specifically, these are intended to delay socket shutdown to the point where the socket is no longer in use, as some interactions can cause exceptions to be raised if EOF is received too early.

Lukasa reviewed (changes requested):
So this seems reasonable, but there are some changes in the tests I don't understand. Notes are inline.

hkwi committed:
update test

introduced FrameEncoderMixin helper class.
fix waits for not displaying unwanted EOF errors.

Lukasa commented on test/test_hyper.py:
```diff
@@ -1107,6 +1114,7 @@ def test_response_version(self):
 
 
 class TestHTTP20Adapter(object):
+
```
Why did all these blank lines get added?

hkwi replied:
Ah... that's autopep8. I'd revert those.

hkwi committed:
revert autopep8 blank lines

Lukasa commented on test/test_integration.py:
```diff
@@ -64,17 +65,30 @@ def frame_buffer():
+    client_preface = b'PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n'
+    timeout = time.time() + 5
+    got = b''
+    while len(got) < len(client_preface) and time.time() < timeout:
```
What's the logic behind this timeout here? The socket doesn't actually have the timeout applied to it, so it could time out indefinitely.

hkwi replied:
Oh, that's a bug. I used timeout for local testing.

Lukasa reviewed (changes requested):
Ok, cool, getting closer. One more question inline.

hkwi committed:
fix socket.recv usage

Lukasa reviewed (approved):
Cool, this looks good to me. Thanks for doing all this work @hkwi! :sparkles:
